### PR TITLE
Enhanced the efficiency of power calculations

### PR DESCRIPTION
**Pull request by haru-44**
#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/combinatorics/free_groups.py [hunk 1]
```diff
@@ -487,19 +487,21 @@ def __str__(self):
...
         if n < 0:
             n = -n
-            return (self.inverse())**n
-
-        result = self
-        for i in range(n - 1):
-            result = result*self
-        # this method can be improved instead of just returning the
-        # multiplication of elements
+            x = self.inverse()
```
Comment (smichr) [inline sympy/combinatorics/free_groups.py:495]: Is it possible to just use `pow(x, n)`?

Comment (haru-44) [reply sympy/combinatorics/free_groups.py:495]: Will a `RecursionError` occur?

**Comment by asmeurer** [comment]
This looks reasonable to me, but we should check that all these methods are actually tested with nontrivial powers.

File: sympy/polys/monomials.py [hunk 2]
```diff
@@ -593,18 +593,20 @@ def __truediv__(self, other):
...
-        else:
+        if n < 0:
             raise ValueError("a non-negative integer expected, got %s" % other)
+        result = [0]*len(self)
+        if not n:
+            return self.rebuild(result)
+        x = self.exponents
+        while True:
+            if n % 2:
+                result = monomial_mul(result, x)
```
Comment (oscarbenjamin) [inline sympy/polys/monomials.py:604]: Shouldn't this just use `monomial_pow`?

**Comment by oscarbenjamin** [comment]
In some cases iterated multiplication is more efficient than exponentiation by squaring if multiplication is not constant time (e.g. for polynomials).

**Comment by haru-44** [comment]
> In some cases iterated multiplication is more efficient than exponentiation by squaring if multiplication is not constant time (e.g. for polynomials).

Is that the case for small `n`?
For sufficiently large n, multiplication can still be efficient even if it is not constant time. If the computational cost of multiplication is $m(n)$, the cost of iterated multiplication is $O(m(n)n)$, while the cost of exponentiation is $O(m(n) \log n)$.

**Comment by oscarbenjamin** [comment]
The computational cost of multiplication depends on the size of both operands like `M(n1,n2)` and for e.g. polynomials with `n1` and `n2` terms is quadratic like `M(n1,n2) ~ n1*n2` at least if the basic polynomial multiplication algorithm is used. In iterated multiplication one operand stays small but in exponentiation by squaring both operands grow.

Consider something like `(x + y + z + t)^8`. The number of terms is:
```
In [37]: [len((Poly(x+y+z+t)**n).coeffs()) for n in range(1, 8+1)]
Out[37]: [4, 10, 20, 35, 56, 84, 120, 165]
```
In exponentiation by squaring the number of multiplies in the ground domain is:
```
In [38]: 4*4 + 10*10 + 35*35
Out[38]: 1341
```
For iterated multiplication it is:
```
In [45]: 4*4 + 4*10 + 4*20 + 4*35 + 4*56 + 4*84 + 4*120
Out[45]: 1316
```
These numbers are close but raising to the power 16 the difference is greater:
```
In [51]: 4*sum([len((Poly(x+y+z+t)**n).coeffs()) for n in range(1, 16)])
Out[51]: 15500

In [52]: 4*4 + 10*10 + 35*35 +165*165
Out[52]: 28566
```
The more symbols you have and the higher the exponent then the more iterated multiplication is better.

If you have a polynomial $p$ of degree $d$ in $r$ symbols and you want to compute $p^n$ then the iterative method is $O(d^{2r}n^{r+1})$ and exponentiation by squaring is $O(d^{2r}n^{2r})$ assuming constant time coefficient multiplication. The iterative method is asymptotically faster by a factor of $n^{r-1}$.

Counting only the number of coefficient multiplies assumes that multiplication in the coefficient domain is constant which is only true for RR, CC and GF(p). In other domains like ZZ the coefficients grow in size as well as the number of terms compounding the effect that multiplying two large operands is slower than multiplying a small operand by a larger one.

A more complete analysis is in e.g. 

> [HEI72] L. E. HEINDEL, "Computation of Powers of Multivariate Polynomials over the
Integers," J. Comput. System Sci. 1 (1972), pp. 1-8. 

https://www.sciencedirect.com/science/article/pii/S0022000072800370

There are also faster algorithms that are more complicated but if we want to stick to simple algorithms then we should remember that iterated multiplication is not necessarily slower than exponentiation by squaring.

**Comment by haru-44** [comment]
Thanks for the explanation. I lacked a thorough understanding of multivariate polynomials.

**Comment by oscarbenjamin** [comment]
Looks good. Thanks

**Pull request merged**